source/_posts/2018-01-12-pager-duty-and-confluence.markdown

```diff
@@ -0,0 +1,256 @@
+  pd_key = File.read(format('%s/.pager_duty/api_key', ENV['HOME']))
+  pdc = PDClient.new( pd_key )
+  pdc.services['services'].each do |service|
+    service_events = pdc.service_events( service['id'] )
+
+    report[service['name']] = { cnt: service_events['incidents'].size, total_time: 0.0, avg_time: 0.0 }
+
+    service_events['incidents'].each do |i|
+      created_at = Time.parse(i['created_at'])
+      resolved_at = Time.parse(i['last_status_change_at'])
+      ticket_time = resolved_at.to_i - created_at.to_i
+
+      report[service['name']][:total_time] += ticket_time
+    end
+
+    if report[service['name']][:cnt] > 0
+      report[service['name']][:avg_time] = (report[service['name']][:total_time] / report[service['name']][:cnt])
+    end
+  end
+
+  report_by_incidents = report.sort{|a,b| b[1][:cnt] <=> a[1][:cnt] }
+  report_by_ticket_time = report.sort{|a,b| b[1][:avg_time] <=> a[1][:avg_time] }
+
+  jira_client = JiraClient.new
+  jira_page = jira_client.get_page('~bryan.kroger', 'PagerDuty report 2017-12' )
+
+  page = ''
+  page << '<ac:structured-macro ac:macro-id="c3c570b3-e820-4787-9242-05af79a10c55" ac:name="chart" ac:schema-version="1">'
+  page << '<ac:parameter ac:name="3D">true</ac:parameter>'
+  page << '<ac:parameter ac:name="width">500</ac:parameter>'
+  page << '<ac:parameter ac:name="title">Pager duty alerts by service.</ac:parameter>'
+  page << '<ac:parameter ac:name="legend">false</ac:parameter>'
+  page << '<ac:parameter ac:name="type">bar</ac:parameter>'
+  page << '<ac:parameter ac:name="height">350</ac:parameter>'
+  page << '<ac:parameter ac:name="dataDisplay">after</ac:parameter>'
+  page << '<ac:rich-text-body>'
+  page << '<table> <tbody> <tr> <th> <p>Service name</p> </th> <th> <p>Count</p> </th> </tr>'
+  report_by_incidents.each do |name, info|
+    page << format('<tr><td>%s</td><td>%s</td></tr>', name, info[:cnt])
+  end
+  page << '</tbody></table>'
+  page << '</ac:rich-text-body>'
+  page << '</ac:structured-macro>'
+
+  page << '<ac:structured-macro ac:macro-id="c3c570b3-e820-4787-9242-05af79a10c55" ac:name="chart" ac:schema-version="1">'
+  page << '<ac:parameter ac:name="3D">false</ac:parameter>'
+  page << '<ac:parameter ac:name="width">500</ac:parameter>'
+  page << '<ac:parameter ac:name="title">Average resolution time by service.</ac:parameter>'
+  page << '<ac:parameter ac:name="legend">false</ac:parameter>'
+  page << '<ac:parameter ac:name="type">bar</ac:parameter>'
+  page << '<ac:parameter ac:name="height">350</ac:parameter>'
+  page << '<ac:parameter ac:name="dataDisplay">after</ac:parameter>'
+  page << '<ac:rich-text-body>'
+  page << '<table> <tbody> <tr> <th> <p>Service name</p> </th> <th> <p>Average resolution time</p> </th> </tr>'
+  report_by_ticket_time.each do |name, info|
+    page << format('<tr><td>%s</td><td>%s</td></tr>', name, pdc.sec_to_human(info[:avg_time]))
+  end
+  page << '</tbody></table>'
+  page << '</ac:rich-text-body>'
+  page << '</ac:structured-macro>'
```

Comment:
Oof. Would this look better as an `Array#join`?

Reply:
I mean, yeah, sure we probably want code to look all pretty and stuff.  And yes, there probably is a more elegant way of doing this.  I keep going back to the idea that this is "just ops" code and doesn't really need the polish.  On the other hand, I'm also defining polish as "properly written" so...there's that.  I mean, nothing really stopping us from doing a quick and dirty hackathon this weekend and making this little sucker shine. ;)